packages/scrapers-v2/src/lexer/lexer.ts

```diff
@@ -0,0 +1,412 @@
+import { assertUnreachable } from "@graduate/common";
+import axios from "axios";
+import * as cheerio from "cheerio";
+import {
+  COURSE_REGEX,
+  HEADER_REGEX,
+  RANGE_BOUNDED,
+  RANGE_LOWER_BOUNDED_MAYBE_EXCEPTIONS_1,
+  RANGE_LOWER_BOUNDED_MAYBE_EXCEPTIONS_2,
+  RANGE_LOWER_BOUNDED_PARSE,
+  RANGE_UNBOUNDED,
+  SUBJECT_REGEX,
+} from "./constants";
+import {
+  CourseRow,
+  HDocument,
+  HRow,
+  HRowType,
+  HSection,
+  MultiCourseRow,
+  RangeBoundedRow,
+  RangeLowerBoundedRow,
+  RangeUnboundedRow,
+  TextRow,
+  WithExceptions,
+} from "./types";
+
+const loadCatalogHTML = async (url: string): Promise<CheerioStatic> => {
+  try {
+    const { data } = await axios.get(url);
+    return cheerio.load(data);
+  } catch (error) {
+    throw error;
+  }
+};
+
+export const scrapeMajorDataFromCatalog = async (
+  url: string
+): Promise<HDocument> => {
+  try {
+    const $ = await loadCatalogHTML(url);
+    // step 1: Transform scraped table into intermediate representation (IR)
+    return transformMajorDataFromCatalog($);
+  } catch (error) {
+    throw error;
+  }
+};
+
+const transformMajorDataFromCatalog = async (
+  $: CheerioStatic
+): Promise<HDocument> => {
+  const majorName: string = parseText($("#site-title").find("h1"));
+  const catalogYear: string = parseText($("#edition")).split(" ")[0];
+  const yearVersion: number = parseInt(catalogYear.split("-")[0]);
+
+  const requirementsContainer = $("#programrequirementstextcontainer");
+  const courseLists = combineAllCourseListTables($, requirementsContainer);
+
+  const prgramRequiredHeading = requirementsContainer
+    .find("h2")
+    .filter(
+      (_idx: number, element: CheerioElement) =>
+        parseText($(element)).includes("Program") &&
+        parseText($(element)).includes("Requirement")
+    );
+  const prgramRequiredHours =
+    Number(parseText(prgramRequiredHeading.next()).split(/[\s\xa0]+/)[0]) || 0;
+
+  return {
+    programRequiredHours: prgramRequiredHours,
+    yearVersion,
+    majorName,
+    sections: await courseLists,
+  };
+};
+
+const combineAllCourseListTables = async (
+  $: CheerioStatic,
+  requirementsContainer: Cheerio
+): Promise<HSection[]> => {
+  // use a stack to keep track of the course list title and description
+  const descriptions: string[] = [];
+  const courseList: HSection[] = [];
+
+  for (const element of requirementsContainer.children().toArray()) {
+    if (HEADER_REGEX.test(element.name)) {
```

Comment:
what would go in the comments? i actually think the information of what element type it is, and the details of the condition, is helpful in this situation, so that the reader knows the context of what element type the element is in each branch of the if/else-if

Reply:
ahh yea that's a fair point

Reply:
ok, we'll add comments